### DISCUSSION TITLE
Github organisation for Wagtail third-party packages?

### DISCUSSION THREAD

**Discussion by allcaps** [wagtail/wagtail#8288] [Ideas]
The Wagtail core team is exploring the possibility of setting up a Github organisation for housing Wagtail third-party packages. Similar to https://github.com/jazzband. See also https://jazzband.co. Another example is https://github.com/rust-lang-nursery. There are many open-source projects with similar approaches.

The idea is that third-party packages are moved to this Github organisation and maintained by the community. In short: _Wagtail community-driven development and maintenance._

I went ahead and claimed the https://github.com/wagtail-nest organisation.

The benefits:
- Better maintained third-packages
- More opportunities to contribute
- Indication of quality
- No bus factor
- Central place on Github

What I'd like to know:
- Is this a good idea?
- Would you like to contribute and become a member?
- Do you maintain a package that you'd like to move to this new organisation?

**Comment by fabienheureux** [comment]
As said on Slack, I love the name but also the idea.

Something that could be interesting is having a branding / readme layout to help bootstrap a quality package, and to help newcomers to identify a wagtail-nest package. 

Do you think it could fit in the scope of wagtail-nest to have a wagtail's package boilerplate for new wagtail-nest projects to have easy to follow guidelines ?

**Reply by tbrlpld** [reply]
I recently started this discussion regarding branding / design system for the Wagtail ecosystem: https://github.com/wagtail/wagtail/discussions/8208

**Reply by allcaps** [reply]
Yes, I'd like the idea of a quality standard for packages and guidelines for setting up new packages.

We already have:

- https://github.com/wagtail/cookiecutter-wagtail-package
- https://github.com/wagtail/wagtail/wiki/Python-Package-Maintenance-Guidelines

**Reply by fabienheureux** [reply]
I am bit late to the party apparently. Thanks a lot for these great resources ! 

Would you say it should be part wagtail-nest' scope ? Or should it be part of the wagtail's documentation ? Or somewhere else...?

**Reply by allcaps** [reply]
These resources are written for packages in the Wagtail namespace (core team maintained packages). But the guidelines are great and could apply to Wagtail Nest packages too.

Maybe Wagtail Nest has some additional requirements? Jazzband requires a badge and a Jazzband bot as maintainer.
https://github.com/jazzband/wagtailmenus/issues/409

**Reply by hazho** [reply]
> Yes, I'd like the idea of a quality standard for packages and guidelines for setting up new packages.
> 
> We already have:
> 
> * https://github.com/wagtail/cookiecutter-wagtail-package
> * https://github.com/wagtail/wagtail/wiki/Python-Package-Maintenance-Guidelines

it should be under an org.name like this:
[Documentation]([https://github.com/wagtail/Third-party-packs/{{](https://github.com/wagtail/%7B%7B) cookiecutter.project_name_kebab }}/blob/main/README.md)
not directly to wagtail git profile like :
![image](https://user-images.githubusercontent.com/20362375/162115434-33716011-cff9-479e-9a6b-d05d01683a52.png)

**Reply by allcaps** [reply]
I'm not sure if I understand this correctly. Sorry! Are you discussing the location of cookiecutter OR wiki OR wagtail-nest?

Github has a fixed URL structure github.com/OWNER_NAME/REPO_NAME/. Where `owner` is an organisation (eg wagtail) OR a `user` (eg hazho).

The organisation has to be something different than `wagtail` to indicate third-party. Your project is the _first-party_, Wagtail the _second-party_, and all other packages _third-party_.

github.com/wagtail-nest/some-package/ shows that packages are _not owned by Wagtail_.

All other examples have the same strategy.

**Reply by fabienheureux** [reply]
> Jazzband requires a badge and a Jazzband bot as maintainer.  

I don't know about you, but I feel "safe" using a package as soon as I see the jazzband badge.
<img width="77" alt="image" src="https://user-images.githubusercontent.com/1702255/162156873-c56f687a-b1a4-4926-934d-855cded2dd38.png">

Same feeling goes for _awesome_ lists, for example https://github.com/springload/awesome-wagtail#readme 
<img width="149" alt="image" src="https://user-images.githubusercontent.com/1702255/162156806-cb7878d9-9975-4de5-b7da-43dc86e8a9b6.png">

**Reply by zerolab** [reply]
How about <img src="https://img.shields.io/badge/wagtail-nest-teal?logo=wagtail">

https://img.shields.io/badge/wagtail-nest-teal?logo=wagtail

**Reply by allcaps** [reply]
It would be great if all third-party packages have a badge indicating the success or failure of a nightly build against stable and latest Wagtail. 

Bit offtopic, but I would love to have these badges as indicators on https://wagtail.org/packages/ too.

**Reply by chris48s** [reply]
@zerolab has performed a highly effective nerd-snipe on me here :smile:  Some options using Wagtail colour palette..

- ![wagtail nest](https://img.shields.io/badge/wagtail_nest-43B1B0?logo=wagtail&logoColor=black) https://img.shields.io/badge/wagtail_nest-43B1B0?logo=wagtail&logoColor=black
- ![wagtail nest](https://img.shields.io/badge/wagtail_nest-black?logo=wagtail) https://img.shields.io/badge/wagtail_nest-black?logo=wagtail
- ![wagtail nest](https://img.shields.io/badge/wagtail-nest-43B1B0?logo=wagtail) https://img.shields.io/badge/wagtail-nest-43B1B0?logo=wagtail
- ![wagtail nest](https://img.shields.io/badge/wagtail-nest-43B1B0?logo=wagtail&labelColor=black) https://img.shields.io/badge/wagtail-nest-43B1B0?logo=wagtail&labelColor=black
- ![wagtail nest](https://img.shields.io/badge/nest-43B1B0?logo=wagtail&logoColor=black) https://img.shields.io/badge/nest-43B1B0?logo=wagtail&logoColor=black

**Comment by FlipperPA** [comment]
This sounds fantastic. I help maintain three Wagtail packages that are (or at least were) used fairly heavily. My availability goes in waves: there will be months where I help with maintenance and upgrades of these three packages, and other times I'm busy with other commitments. I'd be happy to join the organization to help out where I can, and this would help me feel less guilty about not always having my packages completely up to date!

As an example, here's a PR I made to another org's package recently, which I think would also be a good candidate: https://github.com/labd/wagtailstreamforms/pull/191

I'd recommend we have some kind of test suite standard set up - perhaps through GItHub actions? That'd spur me to finally finish up the test suites for my packages.

**Reply by tomkins** [reply]
A 👍 to having a standard(ish) test suite across the packages, something which can test the latest/upcoming versions of Wagtail could potentially be a huge boost.

One of the downsides of various packages is waiting for compatibility with new releases - something which might be easier if the workload is spread across the organisation.

**Reply by zerolab** [reply]
On nightly tests - we have a guide for adding nightly tests - https://github.com/wagtail/wagtail/wiki/Nightly-testing-of-official-plugins and encourage active maintained packages to make use of. with that in hand it should be quite easy to add a nightly badge

**Comment by thibaudcolas** [comment]
Have we considered if Jazzband might be an option for those packages? It takes quite a bit of effort still to keep track of who has access to what, and whether there are enough releasers for given packages within this type of community-run maintenance, which Jazzband has codified quite well.

**Reply by allcaps** [reply]
Good point. We should investigate Jazzband as an option. 

I just found that [wagtailmenus](https://jazzband.co/projects/wagtailmenus) is part of Jazzband. It has an [issue to implement Jazzband guidelines](https://github.com/jazzband/wagtailmenus/issues/409). I signed up as a band member. It is a smooth process. Jazzband-bot is a maintainer on PyPi. To create releases via Github actions? Not sure. They have a [roady](https://jazzband.co/roadies) to help people find their place.

Like you say, Jazzband has codified quite well. But I wonder if this level of rules and automation is necessary. I'm also a member of the [collective](https://github.com/orgs/collective/people), it is the opposite. I signed the contributor’s agreement (with a pen on a dead tree), the rest is 100% open.

**Reply by thibaudcolas** [reply]
IMO it is necessary, primarily because the point of Jazzband is to maintain a high level of trust packages there are well maintained. The automation is there so someone who doesn’t have deep expertise of a given project can be productive with basic maintenance tasks (such as making a release).

How does the collective do releases? That’s the hard thing eventually – with Jazzband anyone can be a member, but only a few people have the right to make releases.

**Comment by RentFreeMedia** [comment]
I released a new Wagtail project just today, which depends on a fork of a jazzband project actually.

A fork because the official repo's code requires the site framework migrations to happen before some others and it creates a circular dependency in terms of initial migrations when Wagtail is added to the mix.  It seems that people are still maintaining the fork in question but no one is around to merge the PRs and push to pypi, so we get into forks of forks in that case.

In short I would say it's a good idea, but I would take some time to examine the details of jazzband's experiences in cases like the above and perhaps amend the rules a bit to try to improve.  A requirement that a Wagtail core person have PR merge and pypi permissions perhaps?  In the above scenario?

**Reply by allcaps** [reply]
The idea is that wagtail-nest would be community-driven. The day to day operations should be possible without the involvement of core team members. 

Third-party packages are called this way for a reason. They are _not endorsed_ by the core team. Core team members might contribute to third-party packages. No different than any other contributor.

It is in the best interest of the Wagtail community that third-party packages are stable, housed, developed and maintained. It is a task for the core team to promote and facilitate that. We try to find the best approach and make it happen.

We do need a clear mission statement and rules. If we decide to focus on Jazzband, these things are already decided for us.

**Reply by chris48s** [reply]
> We do need a clear mission statement and rules

:+1: This. Some examples of questions to consider in this light:

- If a maintainer moves their wagtail-somegubbins package to the wagtail nest github org, what do they give up?
- Who has the final say when it comes to decisions about the future direction of the package? ("is adding this feature a good idea?" or "should we make this breaking change?" type decisions rather than "does this patch fix the bug"). If its the original maintainer, how long would a maintainer need to be absent or non-responsive for before others step in and start making those decisions?
- There are certain things you are talking about laying down as minimum standards. Standards are good, but what happens if they are violated after a package is moved under the organisation?
- etc

These aren't impossible questions to answer, but it is worth thinking these things through before the situations arise and documenting the decisions. Then everyone knows what they're getting into up-front, reducing the scope for conflict in future.

**Comment by hazho** [comment]
this is great, but to be honest, why the Pypi doesn't offer such option?
during publishing any package to the PyPi, there should be settings to identify the package's belongingness, and the type, for example:
what framework related? (choosing from an auto generated list of options, based on previous packages, kinda similar to the Tagging)
is it a third-party package? (considering if this not been checked, it is core framework-related)

at the end of publishing process, the PyPi should auto bind the newly published package to the relevant Organisation on github

**Reply by allcaps** [reply]
Publishing to PyPi isn't the problem. Single person maintained third-packages packages are. This is an attempt to house and maintain packages collaboratively. We try to make abandoned packages community-owned. So that they can receive updates. The original author may transfer their package form their Github account to wagtail-nest. And with that allow the community to take care of it.

There is nothing wrong with the PyPi workflow. Publishing to PyPi is just the final step to make the package pip installable. It got little to do with development and maintenance. There are tools in place to link back from PyPi to the location of the source code.

**Reply by allcaps** [reply]
PS: the left column on PyPi shows all you desire. See https://pypi.org/project/wagtail/ . Classifiers are 'tags'. See [this filter](https://pypi.org/search/?c=Framework+%3A%3A+Wagtail).

**Reply by hazho** [reply]
actually, what I meant is:
The option to integrate PyPi with a repo(you mentioned that there are tools for this, but I didn't find yet), which makes the housing an automated process. 
the original repo can still be the one the author author published, but an auto-generated fork to the housing location is necessary

**Reply by hazho** [reply]
a work around option would be to embed the filtered tag (you shared above) into the thir-party packages housing readme or the entire page (using github-Pages).

**Comment by vsalvino** [comment]
To avoid wagtail-nest from becoming a wasteland / dumping ground of abandoned packages, I think there needs to be some strict criteria.

For example:

Criteria for wagtail-nest admins:
* Must be at least 1 paid person, with dedicated time, managing the org and the packages. The "maintainer-in-residence" so to speak. This could be the employee of a company using wagtail with company-sanctioned time allocated. Or it could be a contractor paid by the wagtail org, for 10-20 hours a week or something like that. This is largely how PyPI works (yes they do pay someone to manage that system - that's why it works!)

Criteria for packages to be accepted:
* Must have a certain number of pypi installs per month.
* Must have been published to pypi for a certain length of time (e.g. 1 year, 2 years)
* Must have at least 2 maintainers/users/volunteers who can lend their knowledge and be publicly listed as "owners" in the README. These people would also be responsible for helping the "maintainer-in-residence" onboard the project.

**Comment by jsma** [comment]
Others have already addressed the "who gets to decide what feature is accepted", "who gets to create a release", and other issues but I wanted to articulate a slightly different way of thinking about this from the perspective of a developer who uses external packages in my projects and as someone who occasionally contributes to Wagtail and the ecosystem. I do not have the perspective of a package author or maintainer, have very limited experience with Python packaging and no PyPI experience other than as an installer of software, so bracket the following how you will.

### Problems with the status quo

As an end user, it's a waste of time to discover a project that _appears_ to be useful only to find out the hard way that pip can't install it due to dependency conflicts. Or forgetting to look at the last commit in Github only to later realize I've created a dependency on abandonware. Or maybe it appeared to be under active development when I made it a dependency and later learned that was the last commit ever, so now I'm stuck maintaining a fork.

As a contributor, time spent navigating the idiosyncrasies of each project is time taken away from dealing with the task at hand (fixing a bug), e.g.:

- How to run tests, if they exists at all and if not, what's the best way to introduce them?
- Code formatting: should I use or disable my editor defaults in the absence of project defined configuration? Or introduce formatting that is inconsistent with existing code in my PR? Or first submit a "format with `black`" PR and then base my bug fix branch on that?
- Dealing with the myriad methods of packaging should I need to bump a version number and trying to do things in the spirit of the package as it is vs the package as it should be if it were rewritten from scratch today
- If I manage to slog through all of that, will my bug fix PR just sit there for years?

### Goals and solutions

From my perspective, wagtail-nest needs to address the following:

- For end users: provide a single place to find high quality, actively maintained Wagtail packages that are guaranteed to be compatible with the currently supported versions of Wagtail.
- For contributors: 
  - provide a standardized, and consistent way for the community to help maintain this high quality across all packages without needing to deal with idiosyncrasies and inconsistencies between packages
  - provide a clear path for bug fixes, dependency updates, etc. to get promptly released, no languishing PRs

Jazzband seems to primarily be about the bus factor issue for projects that have wide adoption but the owner has disappeared or just doesn't have the time or energy to review PRs etc. Jazzband doesn't really seem to address goals like the above directly, so I see value in a distinct and focused wagtail-nest organization that raises the bar for Wagtail packages specifically. 

The first point above seems to me to be paramount. Without a high quality package repo as the end product, what's the point of all the other time and labor? The rest of the considerations boil down to how to make development life easier so a team of volunteers can actually maintain this high quality repository of packages. 

To simplify contributions, it would be great if:

- All packages follow best practices from upstream Wagtail
  - When Wagtail changes the supported versions of Python/Django, all other packages should follow in lockstep
  - Wagtail's formatting/linting/pre-commit configuration should be adopted by all other packages where appropriate (obviously if a package does not touch the front-end, they won't need eslint etc.)
- All packages should have their packaging configuration modernized and unified to follow current best practices (pyproject.toml/setup.cfg, no setup.py, but again I'm very much a noob at Python packaging so don't have strong opinions beyond "do what PyPA suggests *today* and move away from deprecated/legacy methods")
- All packages must have tests, ideally using a consistent testing infrastructure (which should in turn simplify whatever Github actions automation)

In other words, there should be as close to zero guesswork and minimal mental overhead when contributing to upstream Wagtail or to these external packages. Once I have a suitable development environment for contributing to Wagtail core and have wrapped my head around the standards of how to do things, it should be a smooth, lateral transition to get to work on other nest packages. 

[cookiecutter-wagtail-package](https://github.com/wagtail/cookiecutter-wagtail-package) is an obvious starting point for establishing best practices, although it looks like it also needs some updates (an open [issue](https://github.com/wagtail/cookiecutter-wagtail-package/issues/21) for adding black, etc.). Once these standards are codified, the next step is ensuring that each package added to the nest meets these standards. In other words, an acceptable nest package should look like it was created using the cookiecutter template, following current best practices.

This discussion focuses on "third-party" packages, but the various packages split across github.com/wagtail/* and github.com/torchbox/* might be excellent starting points (assuming the current owners agree) for dog-fooding this nest concept, e.g. [torchbox/wagtailmedia](https://github.com/torchbox/wagtailmedia), [torchbox/wagtail-footnotes](https://github.com/torchbox/wagtail-footnotes), and [wagtail/wagtail-generic-chooser](https://github.com/wagtail/wagtail-generic-chooser), just to name a few that I personally use. Core devs already have access to move them over and could then open them up to non-Torchbox/Wagtail core contributors. As it stands, there is wide variability in the code quality, existence of tests, code formatting, packaging, etc. despite having the implied aura of being "official". 

Bottom line, I want a high quality, trustworthy package repo _and_ I also want it to be as easy as possible to jump in and contribute, which I'd love to do more of. If the above are addressed, it would make my limited time for OSS that much more impactful. Cheers!
